### ISSUE TITLE
CURLOPT_PIPEWAIT runs all requests sequentially for HTTP 1.0

### ISSUE THREAD

**Issue by jeroen** [curl/curl#15838]
### I did this

Perhaps this is not a bug, but expected, unfortunate behavior:

When making several concurrent requests with `CURLOPT_PIPEWAIT` to a single server that does not do http keepalive (e.g. http/1.0), then all requests run sequentially and curl never makes more than one connection to the server, even if it could. 

This happens because at each request, curl waits for the pending request to reveal if it can be multiplexed, but then forgets about it immediately once the connection gets closed. Over and over.

Below I try to simulate the issue using a test server `slowtest.opencpu.org` which has `keepalive_timeout 0` set:

```c
#include <stdio.h>
#include <string.h>
#include <curl/curl.h>

void add_one(CURLM *multi_handle){
  CURL *http_handle = curl_easy_init();
  curl_easy_setopt(http_handle, CURLOPT_URL, "http://slowtest.opencpu.org/delay/1");
  curl_easy_setopt(http_handle, CURLOPT_VERBOSE, 1L);
  curl_easy_setopt(http_handle, CURLOPT_PIPEWAIT, 1L);
  curl_multi_add_handle(multi_handle, http_handle);
}

int main(void) {
  curl_global_init(CURL_GLOBAL_DEFAULT);
  CURLM *multi_handle = curl_multi_init();
  add_one(multi_handle);
  add_one(multi_handle);
  add_one(multi_handle);
  add_one(multi_handle);
  add_one(multi_handle);

 
  int still_running = 1;
  do {
    CURLMcode mc = curl_multi_perform(multi_handle, &still_running);
    if(!mc)
      mc = curl_multi_poll(multi_handle, NULL, 0, 1000, NULL);
  } while(still_running);
  return 0;
}
```

### I expected the following

I had expected multiple requests to get performed in parallel after the first one.

Perhaps the multi handle could immediately start the remaining queued requests once is becomes clear that the connection is not multiplexable? Or maybe it can mark pending requests that they do not have to wait anymore if the host supports multiplexing, when the first connection to this same host is closed?

### curl/libcurl version

curl 8.7.1

### operating system

MacOS

**Comment by bagder** [comment]
Testing with current master and 8.11.1 on Linux, I cannot reproduce. The first request needs a response, but once that is received the remaining transfers continue in parallel.

**Comment by bagder** [comment]
I can imagine that there is a case where this can still happen:

If we wait on the response from the server to know what to do next and that server closes the connection immediately after response. curl does not cache any HTTP version info about servers to which there is no connection alive left to, so it will then need to restart the check.

**Comment by jeroen** [comment]
Yes that is precisely the http/1.0 scenario described in the bug report.

**Comment by bagder** [comment]
I don't think that's a bug. It's just an unfortunate side-effect of this feature.It would take some new piece of infra to work around if we were to enhance libcurl for this scenario.

**Comment by jeroen** [comment]
Fair enough. I was hoping there may be a shortcut, at the point where it is determined that the initial request does not upgrade to HTTP/2 and hence won't be multiplexing, the remaining connections could get cleared immediately (rather than waiting for the first request to complete and disconnect).

But I understand that if this is not how it works, it is not worth overhauling things.
